### PR TITLE
Fix/consider timeouts during streaming switching

### LINKED ISSUE #3275: Sometimes SDL doesn't consider timeouts during streaming switching
### Bug Report
Sometimes SDL doesn't consider timeouts during streaming switching

##### Occurrence
Sometimes (0.5%)

##### Reproduction Steps
1. Start SDL, HMI, connect Mobile device, register 2 new streaming applications, e.g. 
 - App_1: `PROJECTION`
 - App_2: `NAVIGATION`
2. Deactivate App_1
=> streaming still continue since app has `STREAMABLE` state
3. Activate App_2

##### Expected Behavior
SDL does:
 - send OnHMIStatus(FULL, STREAMABLE) notification to App_2
 - send OnHMIStatus(LIMITED, NOT_STREAMABLE) notification to App_1
 - send Navi.OnVideoDataStreaming(false) notification to HMI
 - start 1st 'StopStreamingTimeout' timeout

Once timeout is expired SDL does:
 - send Navi.StopStream(App_1) request to HMI
 - send EndService(VIDEO) request to App_1
 - start 2nd 'StopStreamingTimeout' timeout

##### Observed Behavior
SDL does not consider 'StopStreamingTimeout' and send all the messages at once

##### OS & Version Information
* OS/Version: Ubuntu 18.04
* SDL Core Version: https://github.com/smartdevicelink/sdl_core/tree/develop (912e3cf786)
* Testing Against: ATF

##### Test Case, Sample Code, and / or Example App
 - Script: [3139_02_Stop_streaming_gracefully_2nd_timer](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/6_1/3139_02_Stop_streaming_gracefully_2nd_timer.lua)
 - Report and Logs: [2020-03-04_13-10-27.tar.gz](https://github.com/smartdevicelink/sdl_core/files/4288989/2020-03-04_13-10-27.tar.gz)

### PR DESCRIPTION
**Pull request by SNiukalov**
Fixes #3275 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
ATF Script: [3139_02_Stop_streaming_gracefully_2nd_timer](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/6_1/3139_02_Stop_streaming_gracefully_2nd_timer.lua)

### Summary
The issue is happening, when SDL receives audio/video streaming data during changes HMI state (e.g. STREAMABLE -> NOT_STREAMABLE) internally. Earlier SDL considered only HMI_Level, but now after fix: [Fix SDL app stream processing with non-streamable](https://github.com/smartdevicelink/sdl_core/pull/3151) streaming state is also taken into account.
Relation to this change SDL need to consider audio/video application state for sends EndService from ForbidStreaming, which changing according to  timeout for mobile to stop streaming.

### Changelog
* Consider video and audio status before send EndService 
* Update UTs 

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 1]
```diff
@@ -837,8 +837,10 @@ class ApplicationManagerImpl
   /**
    * @brief ForbidStreaming forbid the stream over the certain application.
    * @param app_id the application's id which should stop streaming.
+   * @param service_type Service type to check
```
Comment (ydementieiev) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @SNiukalov maybe remove description here?

Comment (SNiukalov) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @ydementieiev 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/51915151492980976de1ad212161d71d52e6bbfc

File: src/components/application_manager/src/application_manager_impl.cc [hunk 2]
```diff
@@ -3507,6 +3509,21 @@ void ApplicationManagerImpl::ForbidStreaming(uint32_t app_id) {
     UnregisterApplication(app_id, ABORTED);
     return;
   }
+
+  switch (service_type) {
+    case ServiceType::kMobileNav:
+      if (app->video_streaming_allowed()) {
+        return;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/application_manager_impl.cc]: @SNiukalov suggest to add some logging before these returns just for understanding why function was executed in that way

File: src/components/application_manager/src/application_manager_impl.cc [hunk 3]
```diff
@@ -3507,6 +3509,21 @@ void ApplicationManagerImpl::ForbidStreaming(uint32_t app_id) {
     UnregisterApplication(app_id, ABORTED);
     return;
   }
+
+  switch (service_type) {
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/application_manager_impl.cc]: @SNiukalov 
```suggestion
if (ServiceType::kMobileNav == service_type && app->video_streaming_allowed()) {
 LOG("Video streaming is still allowed");
 return;
}

if (ServiceType::kAudio == service_type && app->audio_streaming_allowed()) {
 LOG("Audio streaming is still allowed");
 return;
}

```

File: src/components/media_manager/src/media_manager_impl.cc [hunk 4]
```diff
@@ -313,7 +313,7 @@ void MediaManagerImpl::OnMessageReceived(
   }
 
   if (!application_manager_.CanAppStream(streaming_app_id, service_type)) {
-    application_manager_.ForbidStreaming(streaming_app_id);
+    application_manager_.ForbidStreaming(streaming_app_id, service_type);
     LOG4CXX_ERROR(logger_,
                   "The application trying to stream when it should not.");
```
Comment (AKalinich-Luxoft) [inline src/components/media_manager/src/media_manager_impl.cc]: @SNiukalov could you pls update this message and include service_type here as well?

Comment (SNiukalov) [reply src/components/media_manager/src/media_manager_impl.cc]: @AKalinich-Luxoft 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/cde8652542fc0a0e431ae0a73e877022041d35da

File: src/components/application_manager/src/application_manager_impl.cc (hunk 3 above)
Comment (SNiukalov) [reply src/components/application_manager/src/application_manager_impl.cc]: @AKalinich-Luxoft 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/cde8652542fc0a0e431ae0a73e877022041d35da

File: src/components/application_manager/src/application_manager_impl.cc (hunk 2 above)
Comment (SNiukalov) [reply src/components/application_manager/src/application_manager_impl.cc]: @AKalinich-Luxoft 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/cde8652542fc0a0e431ae0a73e877022041d35da

File: src/components/media_manager/src/media_manager_impl.cc [hunk 5]
```diff
@@ -313,9 +313,11 @@ void MediaManagerImpl::OnMessageReceived(
   }
 
   if (!application_manager_.CanAppStream(streaming_app_id, service_type)) {
-    application_manager_.ForbidStreaming(streaming_app_id);
+    application_manager_.ForbidStreaming(streaming_app_id, service_type);
     LOG4CXX_ERROR(logger_,
-                  "The application trying to stream when it should not.");
+                  "The application trying to stream when it should not."
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/media_manager/src/media_manager_impl.cc]: @jacobkeeler 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/8087167d9b220da58d32f40c0e5940c75376e8fc

File: src/components/application_manager/src/application_manager_impl.cc [hunk 6]
```diff
@@ -3507,6 +3509,18 @@ void ApplicationManagerImpl::ForbidStreaming(uint32_t app_id) {
     UnregisterApplication(app_id, ABORTED);
     return;
   }
+
+  if (ServiceType::kMobileNav == service_type &&
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/src/application_manager_impl.cc]: @jacobkeeler 
This change relate to sends EndService, so I would like them to be as close to this function as possible.
Should I move these changes above the unregister block?

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 7]
```diff
@@ -834,11 +834,8 @@ class ApplicationManagerImpl
    */
   void EndNaviServices(uint32_t app_id) OVERRIDE;
 
-  /**
-   * @brief ForbidStreaming forbid the stream over the certain application.
-   * @param app_id the application's id which should stop streaming.
-   */
-  void ForbidStreaming(uint32_t app_id) OVERRIDE;
+  void ForbidStreaming(uint32_t app_id,
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @jacobkeeler 
Thanks, fixed in:https://github.com/smartdevicelink/sdl_core/pull/3293/commits/525e820749fcb7fdc09579a2ac8d663fe32ef40b

**Pull request merged**